### PR TITLE
Support signing and storing EdDSA key and certificate in java keystore.

### PR DESCRIPTION
**Pull request by patrickkuo**

### CODE REVIEW

**Comment by kchalkias** [comment]
Nice work, the approach seems reasonable and solves the cross-scheme signing issues. I guess some operations should move to Crypto.kt, especially after https://github.com/corda/corda/pull/599 is approved so it will simplify some calls, minimise code duplication (i.e for signing) and get rid of registering providers multiple times (actually no system registering is anymore required).

File: core/src/main/kotlin/net/corda/core/crypto/ContentSignerBuilder.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import java.security.PrivateKey
+import java.security.SecureRandom
+import java.security.Signature
+
+/**
+ *  Provide extra OID look up for signature algorithm not supported by bouncy castle.
+ *  This builder will use bouncy castle's JcaContentSignerBuilder as fallback for unknown algorithm.
+ */
+class ContentSignerBuilder(val signatureAlgorithm: String) {
+    companion object {
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/crypto/ContentSignerBuilder.kt]: Maybe make this an object and add extra param to build as we don't actually use the instance state.

File: core/src/main/kotlin/net/corda/core/crypto/ContentSignerBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        private val additionalAlgorithm: Map<String, AlgorithmIdentifier> = mapOf(
+                EdDSAEngine.SIGNATURE_ALGORITHM to AlgorithmIdentifier(ASN1ObjectIdentifier("1.3.101.112"))
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/crypto/ContentSignerBuilder.kt]: plural additionalAlgorithms?

File: core/src/main/kotlin/net/corda/core/crypto/X509Utilities.kt [hunk 3]
```diff
@@ -58,7 +61,26 @@ object X509Utilities {
...
     init {
-        Security.addProvider(BouncyCastleProvider()) // register Bouncy Castle Crypto Provider required to sign certificates
+        BouncyCastleProvider().apply {
+            // Add EdDSA to bouncy castle security provider.
+            addAlgorithm("KeyFactory.EdDSA", "net.i2p.crypto.eddsa.KeyFactory")
+            addAlgorithm("KeyPairGenerator.EdDSA", "net.i2p.crypto.eddsa.KeyPairGenerator")
+            addAlgorithm("Signature.NONEwithEdDSA", "net.i2p.crypto.eddsa.EdDSAEngine")
+
+            addAlgorithm("Alg.Alias.KeyFactory.1.3.101.112", EdDSAKey.KEY_ALGORITHM)
+            addAlgorithm("Alg.Alias.KeyFactory.OID.1.3.101.112", EdDSAKey.KEY_ALGORITHM)
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/crypto/X509Utilities.kt]: Put comment with link for RFC source of the magic OIDs

**Comment by patrickkuo** [comment]
I have did a few refactoring on X509Utilities to make use of Crypto object for most of the signing and certificate creation. There are quite a lot of new changes, please review again.

File: core/src/main/kotlin/net/corda/core/crypto/Crypto.kt [hunk 4]
```diff
@@ -119,6 +135,25 @@ object Crypto {
...
+            "BCPQC" to BouncyCastlePQCProvider()) // unfortunately, provider's name is not final in BouncyCastlePQCProvider, so we explicitly set it.
+
+    private fun getBouncyCastleProvider() = BouncyCastleProvider().apply {
+        putAll(EdDSASecurityProvider())
+        addKeyInfoConverter(EDDSA_ED25519_SHA512.signatureOID, KeyInfoConverter(EDDSA_ED25519_SHA512))
+    }
+
+    init {
+        // This registration is needed for reading back EdDSA key from java keystore.
+        // TODO: Find a way to make JSK work with bouncy castle provider or implement our own provide so we don't have to register bouncy castle provider.
```
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/Crypto.kt]: small typo "JKS"|

File: core/src/main/kotlin/net/corda/core/crypto/X509Utilities.kt [hunk 5]
```diff
@@ -1,53 +1,32 @@
...
 object X509Utilities {
-
-    val SIGNATURE_ALGORITHM = "SHA256withECDSA"
-    val KEY_GENERATION_ALGORITHM = "ECDSA"
-    // TLS implementations only support standard SEC2 curves, although internally Corda uses newer EDDSA keys.
-    // Also browsers like Chrome don't seem to support the secp256k1, only the secp256r1 curve.
-    val ECDSA_CURVE = "secp256r1"
-
-    val KEYSTORE_TYPE = "JKS"
+    val DEFAULT_TLS_SIGNATURE_ALGORITHMS = Crypto.ECDSA_SECP256R1_SHA256
```
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/X509Utilities.kt]: better rename this to DEFAULT_TLS_SIGNATURE_SCHEME, cause plural (algorithms) may be confusing.

File: core/src/main/kotlin/net/corda/core/crypto/KeyStoreUtilities.kt [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    if (containsAlias(alias)) {
+        this.deleteEntry(alias)
+    }
+    this.setCertificateEntry(alias, cert)
+}
+
+
+/**
+ * Helper method save KeyStore to storage
+ * @param keyStore the KeyStore to persist
```
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/KeyStoreUtilities.kt]: this I guess is not used any more (there are some other comment in this file param references for unused keyStore)

File: core/src/main/kotlin/net/corda/core/crypto/KeyStoreUtilities.kt [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    val keyPass = keyPassword.toCharArray()
+    val key = getKey(alias, keyPass) as PrivateKey
+    val cert = getCertificate(alias) as X509Certificate
+    return CertificateAndKey(cert, KeyPair(cert.publicKey, key))
+}
+
+/**
+ * Extract public X509 certificate from a KeyStore file assuming storage alias is know
+ * @param keyStoreFilePath Path to load KeyStore from
+ * @param storePassword Password to unlock the KeyStore
```
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/KeyStoreUtilities.kt]: two unused @params

File: core/src/main/kotlin/net/corda/core/crypto/X509Utilities.kt [hunk 8]
```diff
@@ -72,311 +51,54 @@ object X509Utilities {
...
-     */
-    @Deprecated("Use [createIntermediateCert(X500Name, CACertAndKey)] instead, specifying full legal name")
-    fun createIntermediateCert(commonName: String,
-                               certificateAuthority: CACertAndKey): CACertAndKey
-            = createIntermediateCert(getDevX509Name(commonName), certificateAuthority)
-
-    /**
-     * Create a de novo root intermediate X509 v3 CA cert and KeyPair.
      * @param subject subject of the generated certificate.
      * @param certificateAuthority The Public certificate and KeyPair of the root CA certificate above this used to sign it
```
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/X509Utilities.kt]: unused comment param

File: core/src/main/kotlin/net/corda/core/crypto/X509Utilities.kt [hunk 9]
```diff
@@ -72,311 +51,54 @@ object X509Utilities {
...
-
-        val issuer = subject
-        val serial = BigInteger.valueOf(random63BitValue())
-        val pubKey = keyPair.public
-
-        // Ten year certificate validity
-        // TODO how do we manage certificate expiry, revocation and loss
+    fun createSelfSignedCACert(subject: X500Name, signatureScheme: SignatureScheme = DEFAULT_TLS_SIGNATURE_ALGORITHMS): CertificateAndKey {
+        val keyPair = generateKeyPair(signatureScheme)
         val window = getCertificateValidityWindow(0, 365 * 10)
```
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/X509Utilities.kt]: This 10 years validity range, should be either defined as input or we should add a TODO comment for future development.

**Comment by patrickkuo** [comment]
Thanks @kchalkias, i have addressed all issues.

**Pull request merged**